doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -473,82 +473,112 @@ Induction
-   structurally different :n:`@term`\s of an inductive set are equal. For
-   example, from :g:`(S (S O))=(S O)` we can derive by absurdity any
-   proposition.
-
-   The argument :n:`@term` is assumed to be a proof of a statement of
-   conclusion :n:`@term = @term` with the two terms being elements of an
-   inductive set. To build the proof, the tactic traverses the normal forms
-   [1]_ of the terms looking for a couple of subterms :g:`u` and :g:`w` (:g:`u`
-   subterm of the normal form of :n:`@term` and :g:`w` subterm of the normal
-   form of :n:`@term`), placed at the same positions and whose head symbols are
-   two different constructors. If such a couple of subterms exists, then the
-   proof of the current goal is completed, otherwise the tactic fails.
+.. _equality-inductive_types:
 
-.. note::
-   The syntax :n:`discriminate @ident` can be used to refer to a hypothesis
-   quantified in the goal. In this case, the quantified hypothesis whose name is
-   :n:`@ident` is first introduced in the local context using
-   :n:`intros until @ident`.
+Equality of inductive types
+---------------------------
 
-.. exn:: No primitive equality found.
-   :undocumented:
+We describe in this section some special purpose tactics dealing with
+equality of inductive sets or types. These tactics use the
+equality :g:`eq:forall (A:Type), A -> A -> Prop`, simply written with the infix
+symbol :g:`=`.
 
-.. exn:: Not a discriminable equality.
-   :undocumented:
+:tacn:`discriminate` and :tacn:`injection`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tacn:: discriminate {? @induction_arg }
+
+   .. todo PR Did I drop too much? (i.e. "The argument :n:`@one_term`
+      is assumed to be a proof of a statement of :n:`@term__1 = @term__2`"
+      How would that work with a non-trivial one_term?
+
+   Proves any goal using any hypothesis in the context in the form
+   :n:`@term__1 = @term__2` that
+   states an impossible structural equality for an inductive type.
+   For example, :g:`(S (S O)) = (S O)` is impossible.
+
+   The tactic traverses the normal forms [1]_ of the terms looking for
+   subterms :g:`u` and :g:`w` that are subterms
+   of the normal forms of :n:`@term__1` and :n:`@term__2`, respectively,
+   placed in the same positions and whose head symbols are
+   different constructors. If such subterms are present, the equality is
+   impossible and the current goal is completed.  Otherwise the tactic fails.
+
+   If no :n:`@induction_arg` is provided, the tactic examines each hypothesis
+   for impossible conditions.
+
+   .. exn:: No discriminable equalities.
+      :undocumented:
+
+   :n:`@natural` (in :n:`@induction_arg`)
+     Equivalent to :tacn:`intros` :n:`until @natural; discriminate @ident`,
+     where :n:`@ident` is the identifier for the last introduced hypothesis.
 
-.. tacv:: discriminate @natural
+   :n:`@ident`  (in :n:`@induction_arg`)
+     Equivalent to :n:`intros until @ident; discriminate @ident`.
```

Comment:
This is not fully correct. This is equivalent only in the case where `ident` is not already the name of a hypothesis from the context. So could be fixed like so:
```suggestion
     If :n:`@ident` is not already in the context, this is equivalent to
     :n:`intros until @ident; discriminate @ident`.
```

Reply:
"primitive equality" meaning Leibniz?
I can drop this case if you like.

Reply:
>"primitive equality" meaning Leibniz?

Right.

> I can drop this case if you like.

It depends on whether we keep explaining `induction_arg` in each tactic (in which case, this case is probably useless for this particular tactic) or if we explain in very general terms outside any tactic description (in which case, we should keep this case, which may makes sense for some tactics, e.g. `destruct`).